pkg/assets/create/creater.go

```diff
@@ -0,0 +1,190 @@
+package create
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+
+	"github.com/openshift/library-go/pkg/assets"
+)
+
+type CreateOptions struct {
+	// Filters allows to filter which files we will read from disk.
+	Filters []assets.FileInfoPredicate
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported.
+// Pass the context to indicate how much time you are willing to wait until all resources are created.
+func EnsureManifestsCreated(ctx context.Context, manifestDir string, restConfig *rest.Config, options CreateOptions) error {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	getRESTMapper := func() (meta.RESTMapper, error) {
+		gr, err := restmapper.GetAPIGroupResources(dc)
+		if err != nil {
+			return nil, err
+		}
+		return restmapper.NewDiscoveryRESTMapper(gr), nil
+	}
+
+	mapper, err := getRESTMapper()
+	if err != nil {
+		return err
+	}
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var lastCreateError error
+	err = wait.PollImmediateUntil(500*time.Millisecond, func() (bool, error) {
```

Comment:
Half a second is not a big deal, but if we are worried about flooding the API server it seems like we'd want to sleep between API requests, and not just between `manifests` iterations.  That would be dropping `PollImmediateUntil` with a raw loop:

```go
while len(manifests) > 0 {
  err, refresh := create(ctx, manifests, client, mapper, options)
  ...
}
```

possibly with a sleep inside `create`'s loop.  If you have any manifests at all, it's hard to see the iteration over those manifests completing in less than half a second.

Reply:
I would be concerned with flooding API server, it should handle creating few resources just fine (right @sttts ?)

I would like to keep `wait.PollImmediateUntil` because it support context and timeouts. If we wait after each create call, calculating timeout will be more complex (that we need).

Reply:
> I would like to keep wait.PollImmediateUntil because it support context and timeouts. If we wait after each create call, calculating timeout will be more complex (that we need).

I've pushed git://github.com/wking/openshift-library-go.git mid-create-cancel 8e509de6f1de4b68b59c1f5097b9a61bbcd8a53f with a stab at dropping `PollImmediateUntil` and moving the wait to per-manifest instead of per-`create` call.  If you're comfortable without a per-manifest rate-limit, I'd recommend just dropping the ticker and looping over manifests as fast as we get responses from the server.  Thoughts?

Reply:
> I've pushed git://github.com/wking/openshift-library-go.git mid-create-cancel 8e509de...

Rebased onto your 0069198 with 8e509de -> fd93f4f.

Reply:
@wking the client itself can be configured to do QPS and rate limiting:

https://github.com/openshift/library-go/blob/e0f627346ad39aa32ada2ed1f17d268ed4b0c3f2/vendor/k8s.io/client-go/rest/config.go#L106

If we are concerned about DOSing the API server, we can set this up in the `*rest.Config` the caller pass to this helper? I can drop the ticker to 10ms or so just to avoid hot-loop, but each individual create call can be limited by using these parameters IMHO.

@sttts @deads2k agree?

Reply:
Does the kubeconfig have a QPS setting as well? I don't think this belongs hardcoded here, but should be settable from the outside. If the kubeconfig knows qps, there is nothing to do.

Reply:
I don't think kubeconfig has that option, we always default qps to 5 in https://github.com/kubernetes/kubernetes/blob/4992402fa5bee9f7ac6fe931d101d362a4295cea/staging/src/k8s.io/client-go/rest/config.go#L276-L278